rfd/00XX-agent-census.md

```diff
@@ -0,0 +1,234 @@
+* Agent: A `teleport` process that runs one or more Teleport services (depending on the configuration).
+* PreHog: A microservice used to capture user events across several Teleport tools.
+
+### Implementation Details
+
+This section is divided in the following subsections:
+- [Data tracked](#data-tracked): which data about each agent will be tracked
+- [Data collection](#data-collection): how such data will flow from the agents to PreHog
+- [Data computation](#data-computation): how to compute such data
+
+#### Data tracked
+
+We want to start tracking the following data in PreHog:
+
+1. Teleport version
+2. Teleport access protocols (`ssh`, `kube`, `app`, `db` and `windows_desktop`)
+3. OS (`linux` or `darwin`, as these are the only two OS currently supported)
+4. OS version (e.g. Linux distribution)
+5. Host architecture (e.g. `amd64`)
+6. `glibc` version
+7. [Installation methods](https://goteleport.com/docs/installation/) (Dockerfile, Helm, `install-node.sh` and `*-ad*.ps1` scripts)
+8. Container runtime (e.g. Docker)
+9. Container orchestrator (e.g. Kubernetes)
+10. Cloud environment (e.g. AWS, GCP, Azure)
+
+#### Data collection
+
+Currently, when an agent first starts, the [inventory control system](https://github.com/gravitational/teleport/tree/6f9ad9553a5b5946f57cb35411c598754d3f926b/lib/inventory) sends an [`UpstreamInventoryHello`](https://github.com/gravitational/teleport/blob/6f9ad9553a5b5946f57cb35411c598754d3f926b/api/proto/teleport/legacy/client/proto/authservice.proto#L1936-L1953) message to the auth server.
+This message has the following fields:
+
+```protobuf
+message UpstreamInventoryHello {
+  string Version = 1;
+  string ServerID = 2;
+  repeated string Services = 3 [(gogoproto.casttype) = "github.com/gravitational/teleport/api/types.SystemRole"];
+  string Hostname = 4;
+}
+```
+
+The `Version` field contains the Teleport version, while the `Services` field contains the subset of the system roles that are currently active at the agent.
+
+We will extend this message to contain the remaining agent data that we want to track:
+
+```protobuf
+message UpstreamInventoryHello {
+  // (...)
+  string OS = 5;
+  string OSVersion = 6;
+  string HostArchitecture = 7;
+  string GLibCVersion = 8;
+  repeated string InstallMethods = 9;
+  string ContainerRuntime = 10;
+  string ContainerOrchestrator = 11;
+  string CloudEnvironment = 12;
+}
+```
+
+When the auth server receives an `UpstreamInventoryHello` message, it will take the information in the message and send it to PreHog.
+For this, a new PreHog `AgentMetadataEvent` message will be added (note that only the `UpstreamInventoryHello.Hostname` won't be sent to PreHog as it can contain PII but also because it doesn't seem useful):
+
+```protobuf
+message AgentMetadataEvent {
+  string version = 1;
+  string server_id = 2;
+  repeated TeleportAccessProtocol protocols = 3;
+  string os = 4;
```

Comment:
Consider making os, and any other appropriate fields, enums (like TeleportAccessProtocol)

[Teleport has an enum for OS](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/devicetrust/v1/os_type.proto)

Reply:
I'm thinking that it might be better to keep them as strings.

If the fields are enums (both in `AgentMetadataEvent` and `UpstreamHelloMessage` (__option A__) or just in `AgentMetadataEvent` (__option B__)), it might make adding a new enum field (e.g. for tracking a new installation method) more complex. If the agent is running a version higher than the auth server (which can happen as patch/minor should be compatible), the auth server won't be able to create an `AgentMetadataEvent` that encodes the new info we want to track: in __option A__, the new enum field will be decoded as the default enum value (i.e. unspecified), while in __option B__ the auth server won't know about the new installation method string and so it can't translate between the two message types.

The only workaround here, I think, would be to first release the new translation code, upgrade all cloud tenants to include it, and only then release the code that tracks the new installation method. But this seems unnecessarily complex.

WDYT?